### PR TITLE
35 database integration

### PR DESCRIPTION
**Pull request by AbiHaddon**
Script for DB integration within software. 

NB: removed database_2.0.py file from commit - see 'database second and third draft' commit - this file was the original script before wrapping script in a class. 

Note identified a potential issue:
When checking the DB for the same version, if the unique_panel_id is the same as one in the database getting the following error. 
image
Will need to be fixed - to discuss in meeting 22/01/2024 between @AmyGrimwood @Fr3dMer @AbiHaddon

### CODE REVIEW

File: src/database_3.0.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            return highest_version, json_query
+        else:
+            return None
+
+#retrieve patient ID and associated panel information
+#condition defined to match either patient_id or sample_id
+
+    def retrieve_patient_and_panel_info(self, input_id):
+
+        condition = or_(self.samples.columns.patient_id == input_id, self.samples.columns.sample_id == input_id)
```
Comment (Fr3dMer) [inline src/database_3.0.py:107]: ### REQUEST
- Nice use of _or(), but what if lab has same patient_id and sample_id for different patient. Would this return multiple results for different patients?

File: src/database_3.0.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    def retrieve_highest_version_json(self, r_code):
+        list_of_versions = self.connection.execute(db.select(self.panels.columns.version)
+            .where(self.panels.columns.r_code == r_code)).fetchall()
+
+        if list_of_versions:
+            highest_version = max(item[0] for item in list_of_versions)
+            json_query = self.connection.execute(db.select(self.panels.columns.JSON)
+                .where((self.panels.columns.r_code == r_code) & (self.panels.columns.version == highest_version))).fetchone()
+            return highest_version, json_query
```
Comment (Fr3dMer) [inline src/database_3.0.py:98]: ### QUERY
- Is there much point in returning highest_version if its in json?

File: src/database_3.0.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            ))
+            return True
+        else:
+            print("Panel with the same r_code and version already exists")
+            return False
+
+#Function for returning JSON when internet connection fails
+#use input r_code to select highest panel version
+
+    def retrieve_highest_version_json(self, r_code):
```
Comment (Fr3dMer) [inline src/database_3.0.py:90]: ### REQUST 
- Possible to add another function using the panel_id assigned by panelapp. Something like this:
```python
def retrieve_highest_version_json(self, panel_id ):
        list_of_versions = self.connection.execute(db.select(self.panels.columns.version)
            .where(self.panels.columns.panel_id == panel_id )).fetchall()
```

File: src/database_3.0.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        self.samples = Table('samples', self.metadata,
+                             Column('sample_id', String(255), primary_key=True, nullable=False),
+                             Column('patient_id', String(255), nullable=False),
+                             Column('unique_panel_id', String(255), ForeignKey('panels.unique_panel_id'), nullable=False),
+                             Column('date', Date())
+                             )
+
+        self.panels = Table('panels', self.metadata,
+                            Column('unique_panel_id', String(255), primary_key=True, nullable=False),
+                            Column('r_code', String(255), nullable=False),
```
Comment (Fr3dMer) [inline src/database_3.0.py:35]: ### REQUEST
- Need to add column for panel_id, the unique identifier panelapp uses internally, as some panels dont have an associated rcode (e.g https://nhsgms-panelapp.genomicsengland.co.uk/panels/166/v1.0) so need a way to search for them

File: src/database_3.0.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+# This line creates a MetaData object which is used to store information about the database schema
+# Creates the table object named samples/panels
+# Metadata creates a new meta data object which holds information about the table
+# The Column() function is used to create each column, with the first argument being the name of the columm and
+# the second argument being the data type
+# Here, the create_all method of the metadata object is called with the engine object as an argument
+# This creates the tables in the database
+
+class PanelAppDB:
+    def __init__(self, db_url='sqlite:///panel_app_db.db'):
```
Comment (Fr3dMer) [inline src/database_3.0.py:21]: ### QUERY
- How are we ensuring we aren't overwriting the db every time the script is run?

**Review by Fr3dMer — CHANGES_REQUESTED** [review]
Looks really good, some last finishing touches requested

File: src/database_3.0.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            ))
+        return True
+
+#Function for inserting a new version of a GMS panel
+#Check version of panel present in database
+#return a boolean indicating whether the operation was successful
+
+    def insert_panel_record(self, unique_panel_id, r_code, version, gene_details, JSON, BED_file_GrCH37, BED_file_GrCH38):
+
+        list_of_unique_id_and_version = self.connection.execute(db.select(self.panels.columns.version)
```
Comment (Fr3dMer) [inline src/database_3.0.py:69]: ### QUERY
- Is this a list of unique_id_and_version or just versions

**Comment by Fr3dMer** [comment]
> Script for DB integration within software.
> 
> NB: removed database_2.0.py file from commit - see 'database second and third draft' commit - this file was the original script before wrapping script in a class.
> 
> Note identified a potential issue: When checking the DB for the same version, if the unique_panel_id is the same as one in the database getting the following error. image Will need to be fixed - to discuss in meeting 22/01/2024 between @AmyGrimwood @Fr3dMer @AbiHaddon

Forgot to discuss in chat yesterday. Is this exception not catching this? 
```python
        else:
            print("Panel with the same r_code and version already exists")
            return False
```

**Review by AmyGrimwood — APPROVED** [review]
Some more inline commenting would help with the documentation requirements

File: src/database_3.0.py (hunk 1 above)
Comment (AbiHaddon) [reply src/database_3.0.py:107]: well sample no. normally starts with a unique prefix i.e. EX.... so I would have thought this wouldn't happen... what do you think?

File: src/database_3.0.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            return False
+
+#Function for returning JSON when internet connection fails
+#use input r_code to select highest panel version
+
+    def retrieve_highest_version_json(self, r_code):
+        list_of_versions = self.connection.execute(db.select(self.panels.columns.version)
+            .where(self.panels.columns.r_code == r_code)).fetchall()
+
+        if list_of_versions:
```
[the remark this answers is not in the document]
Comment (AbiHaddon) [reply src/database_3.0.py:94]: Not sure what you mean by type here?

File: src/database_3.0.py (hunk 6 above)
Comment (AbiHaddon) [reply src/database_3.0.py:69]: Not sure what you mean here...??

**Comment by AbiHaddon** [comment]
In response to your previous comment:

> Script for DB integration within software.

>NB: removed database_2.0.py file from commit - see 'database second and third draft' commit - this file was the original script >before wrapping script in a class.

>Note identified a potential issue: When checking the DB for the same version, if the unique_panel_id is the same as one in the >database getting the following error. image Will need to be fixed - to discuss in meeting 22/01/2024 between >@AmyGrimwood @Fr3dMer @AbiHaddon

>Forgot to discuss in chat yesterday. Is this exception not catching this on line 83?

        else:
          print("Panel with the same r_code and version already exists")
          return False

    I have now fixe this so I am getting the above statement.  The code was comparing the version with an entire tuple

**Comment by AbiHaddon** [comment]
> Some more inline commenting would help with the documentation requireme

**Comment by AbiHaddon** [comment]
> Some more inline commenting would help with the documentation requirements

File: src/database_3.0.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Fr3dMer) [reply src/database_3.0.py:94]: Mostly asking for you to add what the if statement is checking

[the remark this answers is not in the document]
Comment (AbiHaddon) [reply src/database_3.0.py:94]: An in a comment?

File: src/database_3.0.py (hunk 1 above)
Comment (Fr3dMer) [reply src/database_3.0.py:107]: In my lab patient id is an int like 123 while sample id is a float like 23.12345 (dont ask, this was decided in 2008 and we're stuck with it). Is it worth assuming the lab we are making this for does not store each id as the same types, like both as ints? This is a nit picky point so happy to ignore now and add as a future feature

File: src/database_3.0.py (hunk 6 above)
Comment (Fr3dMer) [reply src/database_3.0.py:69]: I was just confused about variable name as its called `list_of_unique_id_and_version` but from my understanding the only data going into it is `(self.panels.columns.version)`, may be getting a little lost here so please let me know if I am

File: src/database_3.0.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Fr3dMer) [reply src/database_3.0.py:94]: more like `if (list_of_versions == xyz):`

File: src/database_3.0.py (hunk 1 above)
Comment (AbiHaddon) [reply src/database_3.0.py:107]: I think so and I think it is sensible to add as a future feature. 'TODO' comment added to script

File: src/database_3.0.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (AbiHaddon) [reply src/database_3.0.py:94]: @AbiHaddon @Fr3dMer as discussed in slack meeting 24/01/24, the following comment to be added to script. 'The if statement is checking whether the list_of_versions is not empty. In Python, an empty list evaluates to False in a boolean context, and a non-empty list evaluates to True'

File: src/database_3.0.py (hunk 6 above)
Comment (AbiHaddon) [reply src/database_3.0.py]: No you are right the variable name is confusing.

Resolved issue.

**Comment by AbiHaddon** [comment]
@Fr3dMer will check pep8 compliance, add doc strings, exception handling as part of sprint 5 issue

**Comment by AbiHaddon** [comment]
> Some more inline commenting would help with the documentation requirements

will add commenting as part of sprint 5

**Review by Fr3dMer — APPROVED** [review]

**Pull request merged**